validator/keymanager/remote-web3signer/keymanager.go

```diff
@@ -38,54 +53,327 @@ type SetupConfig struct {
-func (km *Keymanager) FetchValidatingPublicKeys(ctx context.Context) ([][fieldparams.BLSPubkeyLength]byte, error) {
-	if km.publicKeysURL != "" && !km.publicKeysUrlCalled {
-		providedPublicKeys, err := km.client.GetPublicKeys(ctx, km.publicKeysURL)
+	keyFileExists := false
+	if km.keyFilePath != "" {
+		keyFileExists, err = checkFileExists(km.keyFilePath, maxRetries, retryDelay)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not check if file exists")
+		}
+	}
+
+	var ppk []string
+	// load key values
+	if cfg.PublicKeysURL != "" {
+		providedPublicKeys, err := km.client.GetPublicKeys(ctx, cfg.PublicKeysURL)
 		if err != nil {
 			erroredResponsesTotal.Inc()
-			return nil, errors.Wrap(err, fmt.Sprintf("could not get public keys from remote server url: %v", km.publicKeysURL))
+			return nil, errors.Wrapf(err, "could not get public keys from remote server url: %v", cfg.PublicKeysURL)
+		}
+		ppk = providedPublicKeys
+	} else if len(cfg.ProvidedPublicKeys) != 0 {
+		ppk = cfg.ProvidedPublicKeys
+	}
+
+	// use a map to remove duplicates
+	flagLoadedKeys := make(map[string][48]byte)
+
+	// Populate the map with existing keys
+	for _, key := range ppk {
+		decodedKey, err := hexutil.Decode(key)
+		if err != nil {
+			return nil, errors.Wrap(err, "could not decode public key for remote signer")
+		}
+		if len(decodedKey) != fieldparams.BLSPubkeyLength {
+			return nil, fmt.Errorf("invalid public key length: expected %d, got %d", fieldparams.BLSPubkeyLength, len(decodedKey))
+		}
+		flagLoadedKeys[key] = bytesutil.ToBytes48(decodedKey)
+	}
+	km.flagLoadedKeysMap = flagLoadedKeys
+
+	// load from file
+	if keyFileExists {
+		_, fileKeys, err := km.readKeyFile()
+		if err != nil {
+			return nil, errors.Wrap(err, "could not read key file")
 		}
-		// makes sure that if the public keys are deleted the validator does not call URL again.
-		km.publicKeysUrlCalled = true
-		km.providedPublicKeys = providedPublicKeys
+		maps.Copy(fileKeys, flagLoadedKeys)
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(fileKeys)
+		km.lock.Unlock()
+		// create a file watcher
+		go func() {
+			err = km.refreshRemoteKeysFromFileChangesWithRetry(ctx, retryDelay)
+			if err != nil {
+				log.WithError(err).Error("could not refresh remote keys from file changes")
+			}
+		}()
+	} else {
+		km.lock.Lock()
+		km.providedPublicKeys = maps.Values(flagLoadedKeys)
+		km.lock.Unlock()
+	}
+
+	return km, nil
+}
+
+func checkFileExists(path string, retriesRemaining int, retryDelay time.Duration) (bool, error) {
+	rr := retriesRemaining // make a copy to not alter the original
+	if rr == 0 {
+		return false, errors.New("file check retries remaining exceeded")
 	}
+
+	// Check if the directory and file exists, if not create it
+	keyFileExists, err := file.Exists(path, file.Regular)
+	if err != nil {
+		return false, errors.Wrapf(err, "could not check if remote signer persistent keys exists in %s", path)
+	}
+
+	if !keyFileExists {
```

Comment:
This is not tested. Actually only the case `file does not exists` seems not to be tested.